api/src/main/java/uk/gov/dvsa/mot/trade/read/core/MotrReadServiceDatabase.java

```diff
@@ -0,0 +1,237 @@
+import uk.gov.dvsa.mot.vehicle.api.Vehicle;
+import uk.gov.dvsa.mot.vehicle.read.core.VehicleReadService;
+
+import java.text.SimpleDateFormat;
+import java.util.Date;
+import java.util.List;
+
+public class MotrReadServiceDatabase implements MotrReadService {
+
+    private static final SimpleDateFormat SDF_DATE_ISO_8601 = new SimpleDateFormat("yyyy-MM-dd");
+    private static final SimpleDateFormat SDF_YEAR = new SimpleDateFormat("yyyy");
+
+    private final MotTestReadService motTestReadService;
+    private final VehicleReadService vehicleReadService;
+
+    @Inject
+    public MotrReadServiceDatabase(MotTestReadService motTestReadService, VehicleReadService vehicleReadService) {
+
+        this.vehicleReadService = vehicleReadService;
+        this.motTestReadService = motTestReadService;
+    }
+
+    @Override
+    @ProvideDbConnection
+    public MotrResponse getLatestMotTestByRegistration(String registration) {
+
+        List<Vehicle> vehicles = vehicleReadService.findByRegistration(registration);
+        MotrReadServiceDatabase.VehicleAndLatestMot vehicleAndLatestMot = getVehicleAndLatestMotTestPass(vehicles);
+
+        if (vehicleAndLatestMot == null || !vehicleAndLatestMot.hasMotTest()) {
+            return getDvlaVehicleByRegistration(registration);
+        }
+
+        return mapVehicleAndLatestMotToMotrResponse(vehicleAndLatestMot);
+    }
+
+    @Override
+    @ProvideDbConnection
+    public MotrResponse getLatestMotTestByDvlaVehicleId(Integer dvlaVehicleId) {
+
+        List<Vehicle> vehicles = vehicleReadService.findByDvlaVehicleId(dvlaVehicleId);
+        MotrReadServiceDatabase.VehicleAndLatestMot vehicleAndLatestMot = getVehicleAndLatestMotTestPass(vehicles);
+
+        if (vehicleAndLatestMot == null || !vehicleAndLatestMot.hasMotTest()) {
+            return getDvlaVehicleById(dvlaVehicleId);
+        }
+
+        return mapVehicleAndLatestMotToMotrResponse(vehicleAndLatestMot);
+    }
+
+    @Override
+    @ProvideDbConnection
+    public MotrResponse getLatestMotTestByMotTestNumberWithSameRegistrationAndVin(Long motTestNumber) {
+
+        List<Vehicle> vehicles = vehicleReadService.findByMotTestNumberWithSameRegistrationAndVin(motTestNumber);
+        MotrReadServiceDatabase.VehicleAndLatestMot vehicleAndLatestMot = getVehicleAndLatestMotTestPass(vehicles);
+        if (vehicleAndLatestMot == null) {
+            return null;
+        }
+        return mapVehicleAndLatestMotToMotrResponse(vehicleAndLatestMot);
+    }
+
+    private VehicleAndLatestMot getVehicleAndLatestMotTestPass(List<Vehicle> vehicles) {
+
+        if (CollectionUtils.isNullOrEmpty(vehicles)) {
+            return null;
+        }
+
+        Vehicle vehicle = null;
+        MotTest motTest = null;
+
+        for (Vehicle v : vehicles) {
+            MotTest mt = motTestReadService.getLatestMotTestPassByVehicle(v);
+
+            // If the vehicle or MOT hasn't been set yet
```

Comment:
Could the conditions in this if statement be extracted to a method with a descriptive name?

Reply:
Agreed to address in future release.